packages/platforms/accton/arm64/as4564-26p/modules/builds/src/arm64-accton-as4564-26p-fan.c

```diff
@@ -0,0 +1,527 @@
+{
+	long value;
+	int status;
+
+	status = kstrtol(buf, 10, &value);
+	if (status)
+		return status;
+
+	if (!value)
+		return count;
+
+	/* Read current status */
+	mutex_lock(&data->update_lock);
+
+	status = _reset_wtd();
+	if (unlikely(status < 0)) {
+		dev_dbg(dev, "Unable to reset the watchdog timer\n");
+		return status;
+	}
+
+	mutex_unlock(&data->update_lock);
+	return count;
+}
+
+static struct as4224_fan_data *as4224_fan_update_device(struct device *dev)
+{
+	mutex_lock(&data->update_lock);
+
+	if (time_after(jiffies, data->last_updated + HZ + HZ / 2) ||
+		!data->valid) {
+		int i;
+
+		dev_dbg(dev, "Starting as4224_fan update\n");
+		data->valid = 0;
+
+		/* Update fan data
+		 */
+		for (i = 0; i < ARRAY_SIZE(data->reg_val); i++) {
+			int status = as4224_fan_read_value(fan_reg[i]);
+
+			if (status < 0) {
+				data->valid = 0;
+				mutex_unlock(&data->update_lock);
+				dev_dbg(dev, "reg %d, err %d\n",
+						fan_reg[i], status);
+				return data;
+			}
+			else {
+				data->reg_val[i] = status;
+			}
+		}
+
+		data->last_updated = jiffies;
+		data->valid = 1;
+	}
+
+	mutex_unlock(&data->update_lock);
+
+	return data;
+}
+
+static int as4224_fan_probe(struct platform_device *pdev)
+{
+	int status = -1;
+	int i = 0;
+
+	/* Register sysfs hooks */
+	for (i = 0; i < ARRAY_SIZE(fan_group); i++) {
+		/* Register sysfs hooks */
+		status = sysfs_create_group(&pdev->dev.kobj, &fan_group[i]);
+		if (status) {
+			goto exit;
+		}
+	}
+
+	dev_info(&pdev->dev, "device created\n");
+
+	return 0;
+
+exit:
+	for (--i; i >= 0; i--) {
+		sysfs_remove_group(&pdev->dev.kobj, &fan_group[i]);
+	}
+	return status;
+}
+
+static int as4224_fan_remove(struct platform_device *pdev)
+{
+	int i = 0;
+
+	for (i = 0; i < ARRAY_SIZE(fan_group); i++) {
+		sysfs_remove_group(&pdev->dev.kobj, &fan_group[i]);
+	}
+
+	return 0;
+}
+
+static struct platform_driver as4224_fan_driver = {
+	.probe	  = as4224_fan_probe,
+	.remove	 = as4224_fan_remove,
+	.driver	 = {
+		.name   = DRVNAME,
+		.owner  = THIS_MODULE,
+	},
+};
+
+static int __init as4224_fan_init(void)
+{
+	int ret;
+
+	data = kzalloc(sizeof(struct as4224_fan_data), GFP_KERNEL);
+	if (!data) {
+		ret = -ENOMEM;
+		goto alloc_err;
+	}
+
+	mutex_init(&data->update_lock);
+
+	ret = platform_driver_register(&as4224_fan_driver);
+	if (ret < 0)
+		goto dri_reg_err;
+
+	data->pdev = platform_device_register_simple(DRVNAME, -1, NULL, 0);
+	if (IS_ERR(data->pdev)) {
+		ret = PTR_ERR(data->pdev);
+		goto dev_reg_err;
+	}
+
+	return 0;
+
+dev_reg_err:
+	platform_driver_unregister(&as4224_fan_driver);
+dri_reg_err:
+	kfree(data);
+alloc_err:
+	return ret;
+}
+
+static void __exit as4224_fan_exit(void)
+{
+	platform_device_unregister(data->pdev);
+	platform_driver_unregister(&as4224_fan_driver);
+	kfree(data);
+}
+
+module_init(as4224_fan_init);
+module_exit(as4224_fan_exit);
+
+MODULE_AUTHOR("Brandon Chuang <brandon_chuang@accton.com.tw>");
+MODULE_DESCRIPTION("as4224_fan driver");
+MODULE_LICENSE("GPL");
```

Comment:
Did you copy this from somewhere? If so, could you make it one copy commit, and then a commit adding the changes?

Reply:
The driver is different from as4224.
I had revised the string as as4564.